### ISSUE TITLE
Export of xmap (labels + comments + memory metainfo) for Xpeccy emulator

### ISSUE THREAD

**Issue by Volutar** [z00m128/sjasmplus#161]
Hello!
Xpeccy emulator supports labels and comments and memory metainfo (i.e. code/byte/word/ascii)

File format is not textual.

```
<”XMEMMAP ”> //header
<”ramflags”><4 bytes = length><memory_map>
<memory_map> flags x ram_size
	flags( & 0xf0): 0x00=default(code) 0x10=db 0x20=dw 0x30=dw (addr=label) 0x40=db (ascii text) 0x50=code
<”romflags”><4 bytes = length><memory_map>
	<memory_map> flags x rom_size
		[as above]
<”labels  ”><4 bytes = length><text>
	<text>
SSS:NNNNNNNN:label\n
...
<”comments”><4 bytes = length><text>
	<text>
NNNNNNNN:comment\n
...
```

It consists of number of chunks with 8-char header, 4 byte length of chunks and chunk data itself.
For 48k Spectrum memory model ram section still should consist of 8 pages (but only pages 0,2,5 are used, others are empty).
Rom section is unnecessary for export (though it is possible to make rom images?).
128K ram layout is linear (0,1,2,3,4,5,6,7).

SSS - memory block type "RAM"/"ROM".
"\n" between labels/comments is in Unix format (0x0a), utf8 encoding.
NNNNNNNN - is linear address

Emulator page https://github.com/samstyle/Xpeccy
Latest Window build (supporting xmap) with configs/roms/profiles - https://volutar.myds.me/Xpeccy0.6.20211230.zip
command line parameter to import xmap: "--xmap \<file\>"
Or can be loaded/saved via debugger menu:
![image](https://user-images.githubusercontent.com/3478087/147826987-36583bc2-ce14-4f15-8174-8f4c2e26ba6a.png)

Labels can be added by direct entering in address column. ";" as first character makes it a comment. Memory type can be changed by selecting range and choosing type with RMB context menu (View->). 

Currently comments are single-lined but later they meant to be multilined. Each address can have only one comment and only one label (comment+label per address).

**Comment by ped7g** [comment]
hello, I see this is official docs translated from Russian, which helps a lot.

But before reading more into it, I would prefer to have also some "hard facts", ie. some example of asm file in sjasmplus, and some working map filed saved from the emulator with expected values.

Also I see some thing incompatible with sjasmplus:

- the ZX48 pages in sjasmplus are by default 0,1,2,3, not rom,5,2,0, sjasmplus doesn't have rom page (unfortunately, I tried to add it, but it was getting too complicated). So only 128 target makes sense.
- not all types are tracked by sjasmplus this way (DB is not aware if argument was only text or not, and not sure what is the difference between 0x00 and 0x50)

> Each address can have only one comment and only one label (comment+label per address).
and this is quite often not broken by regular asm project, producing several labels per one address, so it depends if such map file would crash the emulator, or the first/last label would win, etc...

Also there're several map file types already produced by sjasmplus, with SLD files being meant as most rich info for any debugger, including source-file locations for every instruction, so I'm not ecstatic about adding another one (with no obvious advantage).

Checking the official Xpeccy documentation just now, I see it should be able to load the [`LABELSLIST`](https://z00m128.github.io/sjasmplus/documentation.html#po_labelslist) map file, which IIRC does include the zx48 pages translation to 5,2,0 mapping, so this should be already usable with current sjasmplus.

**Comment by Volutar** [comment]
`the ZX48 pages in sjasmplus are by default 0,1,2,3, not rom,5,2,0, sjasmplus doesn't have rom page (unfortunately, I tried to add it, but it was getting too complicated). So only 128 target makes sense.`
Bank numbers could be remapped during xmap export procedure if it's 48k.
`not all types are tracked by sjasmplus this way (DB is not aware if argument was only text or not, and not sure what is the difference between 0x00 and 0x50)`
Actually there's no difference between 0x00 and 0x50 besides 0x00 is simply default, and 0x50 is marked during "runtime mapping" as actually executed parts of code.
`and this is quite often broken by regular asm project, producing several labels per one address, so it depends if such map file would crash the emulator, or the first/last label would win, etc...`
If there are multiple labels per one address - it's fine for assembler - each label converted to same address. But there's no way to reverse this action - disassembler can't know which label is going to be used in this particular place or another. So yes, it simply going to use "last/first" one. Of course without crashing. 

I didn't find any exportable format which has memory "mapping" export, so code and data parts are treated differently. I found structures from cspect format useless - they are not about memory data treating.

There was a question about 'A' or 0x61, as argument. It's not an issue, because inner bytes of code instructions are always "code". Data types are only about data sections, bytes/strings/words.
Words/address has only one difference: words are never converted into labels in debugger, and are "grouped" into multiple values, if there are more than 1 adjacent values. While address is meant to be converted into text label, and currently not grouped (though it can be changed).

If life example is really necessary, it going to take some time. Though emulator itself can be used to create xmap file (as reverse engineering tool, with comments and data sections).

**Comment by Volutar** [comment]
[xmaptest.zip](https://github.com/z00m128/sjasmplus/files/7799480/xmaptest.zip)
This is real life example.
In emulator it looks like this:
![image](https://user-images.githubusercontent.com/3478087/147883535-2f91773c-e226-4d77-af78-2c72a3506395.png)

**Comment by ped7g** [comment]
> I didn't find any exportable format which has memory data "mapping", so code and data parts are treated differently. I found structures from cspect format useless - they are not about memory data treating. SLD is just about source code parallel references.

I think you just don't understand all info available in SLD. Or maybe I'm missing what you mean by "data mapping". Also I'm not sure what you mean by code and data parts, as in machine code everything can be both code and data, in some tricky code and size-coding intros even intentionally working as both at the same time... :) So any such distinction is heuristic at best...

Anyway, thanks for the files, I will try to take a look on it and see what is going on and if I understand it (and whether I can generate enough info during assembling).

**Comment by Volutar** [comment]
> Or maybe I'm missing what you mean by "data mapping".

I mean memory mapping. If we look into source .asm code, we clearly see where are assembler instructions, where is string, and where is data (as in example). Of course there are relocatable things, overlays, and such, but when compilation is happening, each generated byte does have its semantics, just by looking into if it's db/dw/ds or such, or assembler instructions.

**Comment by ped7g** [comment]
yes, that's all part of SLD data.
(of course it doesn't cover cases when you produce instruction by `DB opcode` or when you use part of code also as data, but that trickery is not that common in normal code, so in 99% cases the source is simple enough)

**Comment by Volutar** [comment]
Maybe it's not in v1.0 of SLD, but currently it doesn't produce any information about data types. The file from example has given that information:
`sc.asm|49||0|5|30072|F|text
sc.asm|49||0|5|30072|L|,text,,+used
sc.asm|56||0|5|30118|F|table
sc.asm|56||0|5|30118|L|,table,,+used
`
Just labels, and that's all.

**Comment by Volutar** [comment]
Translated documentation for emulator (in English): https://docs.google.com/document/d/1vdFRntEj_dAUjEV_EsaZBP66kX4VIuxJPHR-gCQRnTc/

**Comment by ped7g** [comment]
you are right, SLD doesn't produce information on the "DB vs DW" level detail.
I will have to take a deeper look into assembling code what is possible, but IIRC the `DB 1` vs `DB "abc"` (number vs string) will be very difficult to achieve.

**Comment by Volutar** [comment]
DB #61,#62,#64 vs DB "abc" is almost impossible to determine after compilation (only by value statistic analysis, and seldom correctly for short data fragments), but when source code is processed, it's clear if there's a "" or just value. By absence or presence of ". Same with WORD vs ADDRESS (words which are repleced by labels in dissassembler): labels are always "address" while direct value is always word.

**Comment by Volutar** [comment]
Are there any chances of this to happen anyhow ... soon-ish?

If I understand correctly it will require simple data type flagging with some simple memory map array on 3rd pass.

**Comment by ped7g** [comment]
I didn't look into it in last couple of months, so no idea.. next release will go probably out in 3-5 weeks (I'm waiting for some feedback on the Lua upgrade, to see if there's something else to do, beyond the decimal parsing which was reported right after 1.20.0 release), I may try to take a look again, if there is some low hanging fruit in this ticket and I can improve it a bit.

But affecting labels definition backward from directive/instruction is not so easy with current architecture. Although I just did it first time with the smart SMC labels, so there's some way how to do it, although that doesn't seem like good fit for larger-scale tagging.

I don't know, I remember this being not simple, will need to take again some deep look, what is possible.
